File: src/core/engine/library/objects.ts
/**
 * Drawing objects and grid cells, `stdlib.md` section 14.
 *
 * Neither goes through a channel, and `compiled-program.md` 4.11 says why: both
 * would need an unbounded number of them, and a channel is by definition one
 * value per bar. A line, label or box is an object in the heap that a script
 * creates once and mutates over many bars; a grid is a handle the engine puts
 * in a slot and its cells are a per-bar buffer.
 *
 * **A drawing object rolls back with everything else.** A moving bar that
 * re-executes ten times must leave one box behind rather than ten, and the
 * object is in the heap, so the heap's undo journal already carries it. Section
 * 6.1's list of what a checkpoint holds does not name the drawing roster
 * separately, and it has to be part of it: without that, creation would survive
 * a rollback and the chart would accumulate a copy per tick.
 *
 * **A deleted object stays deleted rather than becoming absent.** OS4005 names
 * the bar it was deleted on, because a script that mutates a stale handle is
 * holding a name for something it already threw away, and answering "absent"
 * would hide that behind a drawing that quietly stopped moving. An *absent*
 * handle is a different fact and does nothing: a setter given `none` is a gap
 * like every other absence reaching a drawing surface (`language.md` 6.7), and
 * it is what the fix for OS4005 asks a script to produce.
 *
 * **An object is anchored to a time and a price and to nothing else.** Neither
 * anchor is read against the bars: an anchor past the newest bar is kept as it
 * was given, so a projected line reaches into the margin, and an anchor inside
 * the history does not move when more history loads and every bar index
 * shifts. A bar index is a position in the data the engine was given
 * (`language.md` 7.2), and an object that stored one would be somewhere else
 * on the same chart tomorrow.
 *
 * **A polyline holds its own path.** The two arrays it is given are read once,
 * at the call, and the points are the object's from then on; `draw.setPoints`
 * is how a path changes. The alternative, keeping the arrays and reading them
 * when the host draws, would mean a `push` to an array the script kept for its
 * own bookkeeping silently redrawing a shape, and two engines could then
 * disagree about when the path was read.
 */
import type { DrawingObject, GridCell, Value } from '../values/index.js';
import { isDrawing, isNumber, reference } from '../values/index.js';
import { entry, refAt, valueAt, wholeAt } from './binding.js';
import type { CallContext, ManifestEntry } from './binding.js';
import { clearArray } from './arrays.js';

/** The drawing object a handle names, refusing one the script already deleted. */
function drawing(ctx: CallContext, args: readonly Value[], index: number): DrawingObject | undefined {
  const handle = refAt(args, index);
  if (handle === null) return undefined;
  const object = ctx.heap.get(handle.id);
  if (object === undefined || !isDrawing(object)) return undefined;
  if (object.deleted) ctx.guard.deleted(ctx.span, object.kind, object.deletedAt);
  ctx.heap.touch(handle.id);
  return object;
}

/**
 * Creates one object, once the ceiling has been asked.
 *
 * The properties arrive as a function rather than as a record so that the
 * ceiling is charged before any of the work: a polyline copies its whole path
 * into the heap, and an engine that built that and then refused to keep it
 * would have spent the memory it was refusing to spend.
 */
function make(
  ctx: CallContext,
  kind: DrawingObject['kind'],
  props: () => Record<string, Value>,
): Value {
  ctx.guard.drawing(ctx.span, ctx.heap.drawingCount());
  const id = ctx.heap.allocate({ kind, props: props(), deleted: false, deletedAt: -1 });
  return reference(id);
}

/** The numbers one path argument holds, or an empty path when it holds none. */
function pointsIn(ctx: CallContext, args: readonly Value[], index: number): readonly Value[] {
  const held = ctx.heap.deref(valueAt(args, index));
  return held !== undefined && held.kind === 'array' ? held.items : [];
}

/**
 * A polyline's path, copied out of the two arrays that describe it.
 *
 * The arrays are paired by index over the longer of them, so a script that
 * supplies more times than prices gets a point with no price rather than a
 * shorter path: absence reaching a drawing surface is a gap the host draws as a
 * break (`language.md` 6.7), and dropping the odd points instead would quietly
 * redraw the shape. A value that is not a number is absent for the same reason.
 *
 * The copies are heap arrays, so they are rolled back, swept and counted like
 * every other object rather than being a second kind of storage.
 */
function path(ctx: CallContext, args: readonly Value[], first: number): Record<string, Value> {
  const times = pointsIn(ctx, args, first);
  const prices = pointsIn(ctx, args, first + 1);
  const length = Math.max(times.length, prices.length);
  const copy = (source: readonly Value[]): Value => {
    const items: Value[] = [];
    for (let i = 0; i < length; i += 1) {
      const one = source[i];
      items.push(one !== undefined && isNumber(one) ? one : null);
    }
    return reference(ctx.heap.allocate({ kind: 'array', items }));
  };
  return { times: copy(times), prices: copy(prices) };
}

/** A mutator that writes named fields onto an existing object. */
function setter(
  name: string,
  params: string,
  fields: (args: readonly Value[]) => Record<string, Value>,
): ManifestEntry {
  return entry(name, params, (ctx, args) => {
    const object = drawing(ctx, args, 0);
    if (object === undefined) return null;
    object.props = { ...object.props, ...fields(args) };
    return null;
  });
}

export const OBJECT_ENTRIES: readonly ManifestEntry[] = [
  entry('draw.line', 't1 p1 t2 p2 color width style extendLeft extendRight', (ctx, args) =>
    make(ctx, 'line', () => ({
      t1: valueAt(args, 0),
      p1: valueAt(args, 1),
      t2: valueAt(args, 2),
      p2: valueAt(args, 3),
      color: valueAt(args, 4),
      width: valueAt(args, 5),
      style: valueAt(args, 6),
      extendLeft: valueAt(args, 7),
      extendRight: valueAt(args, 8),
    })),
  ),

  entry('draw.label', 't p text color textColor align tooltip', (ctx, args) =>
    make(ctx, 'label', () => ({
      t: valueAt(args, 0),
      p: valueAt(args, 1),
      text: valueAt(args, 2),
      color: valueAt(args, 3),
      textColor: valueAt(args, 4),
      align: valueAt(args, 5),
      tooltip: valueAt(args, 6),
    })),
  ),

  entry(
    'draw.box',
    't1 p1 t2 p2 color fillColor opacity width text textColor tooltip',
    (ctx, args) =>
      make(ctx, 'box', () => ({
        t1: valueAt(args, 0),
        p1: valueAt(args, 1),
        t2: valueAt(args, 2),
        p2: valueAt(args, 3),
        color: valueAt(args, 4),
        fillColor: valueAt(args, 5),
        opacity: valueAt(args, 6),
        width: valueAt(args, 7),
        text: valueAt(args, 8),
        textColor: valueAt(args, 9),
        tooltip: valueAt(args, 10),
      })),
  ),

  entry('draw.polyline', 'times prices color width closed fillColor opacity', (ctx, args) =>
    make(ctx, 'polyline', () => ({
      ...path(ctx, args, 0),
      color: valueAt(args, 2),
      width: valueAt(args, 3),
      closed: valueAt(args, 4),
      fillColor: valueAt(args, 5),
      opacity: valueAt(args, 6),
    })),
  ),

  setter('draw.setFrom', 'obj t p', (args) => ({ t1: valueAt(args, 1), p1: valueAt(args, 2) })),
  setter('draw.setTo', 'obj t p', (args) => ({ t2: valueAt(args, 1), p2: valueAt(args, 2) })),
  setter('draw.setAt', 'obj t p', (args) => ({ t: valueAt(args, 1), p: valueAt(args, 2) })),
  setter('draw.setBounds', 'obj t1 p1 t2 p2', (args) => ({
    t1: valueAt(args, 1),
    p1: valueAt(args, 2),
    t2: valueAt(args, 3),
    p2: valueAt(args, 4),
  })),
  entry('draw.setPoints', 'obj times prices', (ctx, args) => {
    const object = drawing(ctx, args, 0);
    if (object === undefined) return null;
    object.props = { ...object.props, ...path(ctx, args, 1) };
    return null;
  }),
  setter('draw.setText', 'obj text', (args) => ({ text: valueAt(args, 1) })),
  setter('draw.setColor', 'obj color', (args) => ({ color: valueAt(args, 1) })),
  setter('draw.setTextColor', 'obj color', (args) => ({ textColor: valueAt(args, 1) })),
  setter('draw.setFillColor', 'obj color', (args) => ({ fillColor: valueAt(args, 1) })),
  setter('draw.setWidth', 'obj width', (args) => ({ width: valueAt(args, 1) })),
  setter('draw.setStyle', 'obj style', (args) => ({ style: valueAt(args, 1) })),
  setter('draw.setExtend', 'obj left right', (args) => ({
    extendLeft: valueAt(args, 1),
    extendRight: valueAt(args, 2),
  })),
  setter('draw.setTooltip', 'obj text', (args) => ({ tooltip: valueAt(args, 1) })),

  // Deleting twice is not an error: the script asked for the object to be gone
  // and it is gone. Only a change to a deleted object is OS4005.
  entry('draw.delete', 'obj', (ctx, args) => {
    const handle = refAt(args, 0);
    if (handle !== null) ctx.heap.deleteDrawing(handle.id, ctx.bar.index);
    return null;
  }),

  entry('draw.deleteAll', '', (ctx) => {
    for (const { id } of ctx.heap.drawings()) ctx.heap.deleteDrawing(id, ctx.bar.index);
    return null;
  }),

  entry('draw.count', '', (ctx) => ctx.heap.drawingCount()),

  /**
   * `cell(t, row, col, text, ...)`: one cell of a grid for this bar.
   *
   * The buffer is emptied at step 3 of every execution, so a re-executed bar
   * rebuilds the whole grid and a checkpoint holds none of it. That is why a
   * grid needs no rollback of its own while a drawing object does.
   */
  entry('cell', 't row col text textColor bgColor align', (ctx, args) => {
    const handle = refAt(args, 0);
    if (handle === null) return null;
    const object = ctx.heap.get(handle.id);
    if (object === undefined || object.kind !== 'table') return null;
    const row = wholeAt(ctx, 'cell', 'row', args, 1);
    const col = wholeAt(ctx, 'cell', 'col', args, 2);
    if (row === null || col === null) return null;
    if (row >= object.rows || col >= object.cols) {
      ctx.guard.badIndex(
        ctx.span,
        ctx.nameOf(valueAt(args, 0)),
        `${row}, ${col}`,
        object.rows * object.cols,
      );
    }
    const written: GridCell = {
      row,
      col,
      text: valueAt(args, 3),
      textColor: valueAt(args, 4),
      bgColor: valueAt(args, 5),
      align: valueAt(args, 6),
    };
    object.cells.push(written);
    return null;
  }),

  entry('clear', 'handle', (ctx, args) => {
    if (clearArray(ctx, args)) return null;
    const handle = refAt(args, 0);
    if (handle === null) return null;
    const object = ctx.heap.get(handle.id);
    if (object === undefined || object.kind !== 'table') return null;
    object.cells = [];
    return null;
  }),
];

File: engine/openscript/objects.py
"""Drawing objects and grids: ``stdlib.md`` sections 14.3 and 14.4.

Neither goes through a channel, and ``compiled-program.md`` 4.11 says why: a
channel is one value per bar, and a script may hold any number of lines, labels
and boxes, created once and changed over many bars, and write any number of
cells into a grid. The calls live here rather than in ``library/`` for the same
reason ``arrays.py`` does: three of them refuse, and the library refuses nothing.

- **A change to a deleted object is OS4005**, naming the bar it was deleted on.
  A script that moves a line it already threw away is holding a name for
  nothing, and answering absence would hide that behind a line that quietly
  stopped moving. An *absent* handle is a different fact and does nothing, which
  is what the fix for OS4005 asks a script to hand over.
- **A cell outside its grid is OS4008**, and a row or a column that is not a
  whole number of zero or more is OS4003 against that argument.
- **The object past the ceiling is OS5010.** The ceiling is the first engine's
  number, for the reason ``budget.EngineLimits`` gives for the string ceiling:
  a case that reaches it must be refused by both engines or by neither.

**A drawing object rolls back with everything else.** A moving bar executed ten
times leaves one box and not ten. The roster is part of the checkpoint (``Run``
takes ``mark`` beside the cells and restores it beside them), and it is cheap to
take because nothing in it is changed in place: a setter replaces an object's
properties with a new mapping, so the mark is a shallow copy of the roster and
the objects it points at are never written again.

**A handle is the object's identity, and a copy of the cells keeps it.** Cells
are copied whole at a checkpoint, so a handle held in a cell would come back as
a second handle to the same object, and ``==`` between two of them would change
its answer across a restore. A handle therefore copies to itself. The same holds
for a grid, whose cells are an output buffer that step 3 empties and nothing
rolls back.

**What a deleted object leaves behind is kept only while a handle names it.**
OS4005 needs the bar an object was deleted on for as long as a script can still
reach it, and not a moment longer: a study that deletes yesterday's line and
draws today's, on every bar, would otherwise keep one record per bar for the
life of the chart and copy all of them at every checkpoint. The record is keyed
weakly by the handle, so it goes when the last name for the object does, which
is also when nothing could ask about it.
"""

import weakref
from typing import Any, Dict, List, Optional, Sequence, Tuple

from .arrays import Refused
from .canonical import canonical_number
from .values import ABSENT, ArrayValue, Reference, is_number, is_whole

#: The kinds ``stdlib.md`` 14.4 creates, each with the property pairs that anchor
#: it, in order. A polyline holds its own path instead, section 14.4's `times`
#: and `prices`, copied at the call.
ANCHORS: Dict[str, Tuple[Tuple[str, str], ...]] = {
    "line": (("t1", "p1"), ("t2", "p2")),
    "box": (("t1", "p1"), ("t2", "p2")),
    "label": (("t", "p"),),
    "polyline": (),
}

#: The two properties a polyline's path is held under.
PATH = ("times", "prices")


class Handle(Reference):
    """One drawing object's identity. Equality is identity, and a copy is itself."""

    __slots__ = ("serial", "kind", "__weakref__")

    def __init__(self, serial: int, kind: str) -> None:
        self.serial = serial
        self.kind = kind

    def __deepcopy__(self, memo: Dict[int, Any]) -> "Handle":
        return self


class Grid(Reference):
    """One declared grid, and the cells the bar being executed wrote into it."""

    __slots__ = ("key", "title", "rows", "cols", "cells")

    def __init__(self, key: str, title: str, rows: int, cols: int) -> None:
        self.key = key
        self.title = title
        self.rows = rows
        self.cols = cols
        self.cells: List[Tuple[Any, ...]] = []

    def __deepcopy__(self, memo: Dict[int, Any]) -> "Grid":
        return self


class Objects:
    """The drawing roster and the declared grids of one run."""

    def __init__(self, ceiling: int, grids: Sequence[Tuple[int, Grid]] = ()) -> None:
        self.ceiling = ceiling
        #: The objects the script holds, by serial, in creation order.
        self.live: Dict[int, Tuple[Handle, Dict[str, Any]]] = {}
        #: The bar each deleted object was deleted on, while a handle names it.
        self.deleted: "weakref.WeakKeyDictionary[Handle, int]" = weakref.WeakKeyDictionary()
        self.serial = 0
        #: Each grid with the slot step 5 puts it in.
        self.grids: List[Tuple[int, Grid]] = list(grids)

    # -- the checkpoint --------------------------------------------------------

    def mark(self) -> Tuple[Any, ...]:
        return (dict(self.live), self.deleted.copy(), self.serial)

    def restore(self, mark: Tuple[Any, ...]) -> None:
        live, deleted, serial = mark
        self.live = dict(live)
        self.deleted = deleted.copy()
        self.serial = serial

    # -- steps 3 and 5 ---------------------------------------------------------

    def clear_grids(self) -> None:
        for _, grid in self.grids:
            grid.cells.clear()

    # -- what a host is handed ------------------------------------------------

    def drawings(self) -> List[Tuple[str, List[Tuple[Any, Any]], Dict[str, Any]]]:
        """Every object the script holds, oldest first: kind, anchors, and the rest."""
        out = []
        for handle, props in self.live.values():
            out.append((handle.kind, _anchors(handle.kind, props), _style(handle.kind, props)))
        return out

    # -- the calls ---------------------------------------------------------------

    def create(self, kind: str, props: Dict[str, Any]) -> Handle:
        if len(self.live) >= self.ceiling:
            raise Refused("OS5010", max=self.ceiling, found=len(self.live) + 1)
        handle = Handle(self.serial, kind)
        self.serial += 1
        self.live[handle.serial] = (handle, props)
        return handle

    def change(self, held: Any, bar: int, fields: Dict[str, Any]) -> Any:
        """A setter: new fields onto a live object, OS4005 onto a deleted one."""
        if not isinstance(held, Handle):
            return ABSENT
        gone = self.deleted.get(held)
        if gone is not None:
            raise Refused("OS4005", kind=held.kind, bar=gone)
        found = self.live.get(held.serial)
        if found is not None:
            self.live[held.serial] = (held, {**found[1], **fields})
        return ABSENT

    def delete(self, held: Any, bar: int) -> Any:
        """Deleting twice is not an error: the object is gone either way."""
        if isinstance(held, Handle) and self.live.pop(held.serial, None) is not None:
            self.deleted[held] = bar
        return ABSENT


def _number(value: Any) -> Any:
    return value if is_number(value) else ABSENT


def _anchors(kind: str, props: Dict[str, Any]) -> List[Tuple[Any, Any]]:
    if kind == "polyline":
        times, prices = props.get(PATH[0], ()), props.get(PATH[1], ())
        return [(_number(times[i]), _number(prices[i])) for i in range(len(times))]
    return [(_number(props.get(time)), _number(props.get(price))) for time, price in ANCHORS[kind]]


def _style(kind: str, props: Dict[str, Any]) -> Dict[str, Any]:
    anchored = set(PATH)
    for time, price in ANCHORS[kind]:
        anchored.update((time, price))
    return {name: value for name, value in props.items() if name not in anchored}


def _at(arguments: Sequence[Any], at: int) -> Any:
    return arguments[at] if at < len(arguments) else ABSENT


def _named(arguments: Sequence[Any], names: str, first: int = 0) -> Dict[str, Any]:
    """The arguments from ``first`` on, under the names the call gives them."""
    return {name: _at(arguments, first + at) for at, name in enumerate(names.split())}


def _path(arguments: Sequence[Any], first: int) -> Dict[str, Any]:
    """A polyline's path, copied out of its two arrays at the call.

    Paired by index over the longer array, so more times than prices is a point
    with no price rather than a shorter path, and a value that is not a number
    is absent: a gap the host draws as a break rather than a shape redrawn.
    """
    held = [_at(arguments, first), _at(arguments, first + 1)]
    items = [one.elements if isinstance(one, ArrayValue) else [] for one in held]
    length = max(len(items[0]), len(items[1]))
    return {
        name: tuple(_number(source[i]) if i < len(source) else ABSENT for i in range(length))
        for name, source in zip(PATH, items)
    }


def _whole(argument: str, value: Any) -> Optional[int]:
    """A row or a column: absent is nothing to write, and anything else not whole is OS4003."""
    if value is ABSENT:
        return None
    if not is_whole(value) or value < 0:
        raise Refused("OS4003", name="cell", argument=argument, found=_described(value))
    return int(value)


def _described(value: Any) -> str:
    if isinstance(value, str):
        return '"' + value + '"'
    if isinstance(value, bool):
        return "true" if value else "false"
    if is_number(value):
        return canonical_number(float(value))
    return "an object"


def _cell(objects: Objects, bar: int, arguments: Sequence[Any]) -> Any:
    grid = _at(arguments, 0)
    if not isinstance(grid, Grid):
        return ABSENT
    row = _whole("row", _at(arguments, 1))
    col = _whole("col", _at(arguments, 2))
    if row is None or col is None:
        return ABSENT
    if row >= grid.rows or col >= grid.cols:
        raise Refused("OS4008", row=row, column=col, rows=grid.rows, columns=grid.cols)
    grid.cells.append((row, col) + tuple(_at(arguments, at) for at in range(3, 7)))
    return ABSENT


def clears_a_grid(arguments: Sequence[Any]) -> bool:
    """``clear(t)`` on a grid: its cells for this bar are gone. False for anything else."""
    grid = _at(arguments, 0)
    if not isinstance(grid, Grid):
        return False
    grid.cells.clear()
    return True


#: The names a cell is written under, the ones ``cell`` gives its arguments.
CELL_FIELDS = ("row", "col", "text", "textColor", "bgColor", "align")


def _made(kind: str, names: str) -> Any:
    return lambda objects, bar, arguments: objects.create(kind, _named(arguments, names))


def _setter(names: str) -> Any:
    return lambda objects, bar, arguments: objects.change(
        _at(arguments, 0), bar, _named(arguments, names, 1)
    )


def _polyline(objects: Objects, bar: int, arguments: Sequence[Any]) -> Any:
    styled = _named(arguments, "color width closed fillColor opacity", 2)
    return objects.create("polyline", {**_path(arguments, 0), **styled})


def _delete_all(objects: Objects, bar: int, arguments: Sequence[Any]) -> Any:
    for handle, _ in list(objects.live.values()):
        objects.delete(handle, bar)
    return ABSENT


#: Every call, by name and argument count, as the manifest keys it. Each takes
#: the run's objects, the bar index and the arguments.
CALLS: Dict[Tuple[str, int], Any] = {
    ("draw.line", 9): _made("line", "t1 p1 t2 p2 color width style extendLeft extendRight"),
    ("draw.label", 7): _made("label", "t p text color textColor align tooltip"),
    ("draw.box", 11): _made("box", "t1 p1 t2 p2 color fillColor opacity width text textColor tooltip"),
    ("draw.polyline", 7): _polyline,
    ("draw.setFrom", 3): _setter("t1 p1"),
    ("draw.setTo", 3): _setter("t2 p2"),
    ("draw.setAt", 3): _setter("t p"),
    ("draw.setBounds", 5): _setter("t1 p1 t2 p2"),
    ("draw.setPoints", 3): lambda objects, bar, arguments: objects.change(
        _at(arguments, 0), bar, _path(arguments, 1)
    ),
    ("draw.setText", 2): _setter("text"),
    ("draw.setColor", 2): _setter("color"),
    ("draw.setTextColor", 2): _setter("textColor"),
    ("draw.setFillColor", 2): _setter("fillColor"),
    ("draw.setWidth", 2): _setter("width"),
    ("draw.setStyle", 2): _setter("style"),
    ("draw.setExtend", 3): _setter("extendLeft extendRight"),
    ("draw.setTooltip", 2): _setter("tooltip"),
    ("draw.delete", 1): lambda objects, bar, arguments: objects.delete(_at(arguments, 0), bar),
    ("draw.deleteAll", 0): _delete_all,
    ("draw.count", 0): lambda objects, bar, arguments: float(len(objects.live)),
    ("cell", 7): _cell,
}
